Adding NodeName to EndpointSlice API, deprecation updates

robscott opened the pull request:
**What type of PR is this?**
/kind feature
/kind api-change

**What this PR does / why we need it**:
In addition to adding NodeName, this notes that the topology field will be deprecated soon. It also removes the IP address type that was deprecated in Kubernetes 1.17 and intended to be removed in 1.20.

Although we were originally scheduled to graduate this API to GA in 1.20 it was [determined that we needed to do a bit more cleanup first](https://github.com/kubernetes/kubernetes/pull/96318#issuecomment-725008391). This PR should accomplish that.

**Does this PR introduce a user-facing change?**:
```release-note
- EndpointSlice has a new NodeName field guarded by the EndpointSliceNodeName feature gate.
- EndpointSlice topology field will be deprecated in an upcoming release.
- EndpointSlice "IP" address type is formally removed after being deprecated in Kubernetes 1.17.
- The discovery.k8s.io/v1alpha1 API is deprecated and will be removed in Kubernetes 1.21.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
Enhancement Issue: https://github.com/kubernetes/enhancements/issues/752
More Context: https://github.com/kubernetes/kubernetes/pull/96318#issuecomment-725008391

/cc @aojea @andrewsykim @bowei @freehan
/assign @thockin
/sig network
/priority important-soon

robscott committed:
Adding NodeName to EndpointSlice API, deprecation updates

In addition to adding NodeName, this notes that the topology field will
be deprecated soon. It also removes the IP address type that was
deprecated in Kubernetes 1.17 and intended to be removed in 1.20.

robscott committed:
Adding EndpointSliceNodeName feature gate

robscott committed:
Updating EndpointSlice controllers to support NodeName field

robscott committed:
Updating ControlPlane to support NodeName field

robscott committed:
Removing "IP" from supported EndpointSlice address types in kube-proxy

robscott commented:
/triage accepted

aojea commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -53,29 +52,44 @@ var (
+	// allow NodeName value if the feature gate is set.
+	allowNodeName := utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName)
+
+	if !allowNodeName {
```
this means that if we have the feature gate disabled for node name, but one of the old slices has the field set, we allow node name?

aojea replied:
ic, there is also a test, I always find this part confusing so take it with a grain of salt, but I thought that the recommended way is to clear the field in strategy and only validate it here is is not nil 
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-field-in-existing-api-version

andrewsykim replied:
> I thought that the recommended way is to clear the field in strategy and only validate it here is is not nil

Agreed, I thought we need to drop disabled fields in strategy, the validation logic here seems fine though. Left a related comment here https://github.com/kubernetes/kubernetes/pull/96440#discussion_r521423193

robscott replied:
Yep, I believe this is intended to help with a scenario where a newer client hits an older API server. I've seen this similar pattern used in several places, including my own [use of it with AppProtocol](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/apis/core/validation/validation.go#L4180-L4187).

robscott replied:
My local copy of this page missed the follow up comments here before I commented - will update strategy here as well, thanks for the references!

robscott replied:
Strategy has been updated.

andrewsykim commented on pkg/apis/discovery/validation/validation_test.go:
```diff
@@ -537,9 +594,10 @@ func TestValidateEndpointSliceUpdate(t *testing.T) {
-		oldEndpointSlice *discovery.EndpointSlice
+		expectedErrors      int
+		nodeNameGateEnabled bool
+		newEndpointSlice    *discovery.EndpointSlice
```
really minor nit, can we put oldEndpointSlice above newEndpointSlice, just makes the test easier to read

robscott replied:
yeah that was really bugging me too, but I wasn't sure I wanted to make the diff that big. I think it is a big enough improvement that I'll add it as a separate commit.

robscott replied:
Updated in latest commit.

andrewsykim commented on pkg/apis/discovery/validation/validation_test.go:
```diff
@@ -552,21 +610,99 @@ func TestValidateEndpointSliceUpdate(t *testing.T) {
+			},
+			expectedErrors: 1,
+		},
+		"node name set after, gate disabled": {
```
This test name should be "node name set before, gate disabled"

robscott replied:
Good catch, updated this to actually do what the name was describing.

andrewsykim commented on pkg/controller/endpointslice/utils.go:
```diff
@@ -87,6 +87,10 @@ func podToEndpoint(pod *corev1.Pod, node *corev1.Node, service *corev1.Service,
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName) {
+		ep.NodeName = &pod.Spec.NodeName
```
If `pod.Spec.NodeName` is empty, should we set `ep.NodeName` to nil instead?

robscott replied:
Updated, thanks!

andrewsykim commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -110,6 +120,17 @@ func validateEndpoints(endpoints []discovery.Endpoint, addrType discovery.Addres
+					allErrs = append(allErrs, field.Invalid(nnPath, *endpoint.NodeName, msg))
+				}
+			} else {
+				allErrs = append(allErrs, field.Forbidden(nnPath, "may not be set unless EndpointSliceNodeName feature gate is enabled"))
```
I didn't see anything in EndpointSlice strategy that would drop this field when the gate is enabled. This means we would reject clients on validation instead of dropping the disable field. Is that expected?

robscott replied:
Yep, that's correct, this relies on validation and will fail instead of dropping the field. This matches the [approach taken for AppProtocol](https://github.com/kubernetes/kubernetes/blob/release-1.18/pkg/apis/core/validation/validation.go#L4180-L4187) and some of the other new fields I've seen. I personally think it's more clear to fail upfront than to silently drop fields. 

I'm not sure if there's a standard written anywhere though, I noticed you're dropping the new `terminating` and `serving` conditions in strategy and have not added anything to prevent them in validation. Either way, it seems like we should probably be using the same approach for both cases. @liggitt or @thockin which approach should we be using here for new fields (dropping in strategy or blocking in validation)?

robscott replied:
Nevermind, @aojea provided a good [reference link](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-field-in-existing-api-version) above for why this should be added to strategy, working on that now.

robscott committed:
Cleaning up EndpointSlice update validation tests

robscott commented:
Tests are all failing due to a strange merge issue. Seems to be affecting a number of PRs and is likely related to https://github.com/kubernetes/test-infra/issues/19910.

robscott commented:
/retest

robscott committed:
Updating EndpointSlice strategy to cover alpha NodeName field

robscott commented:
@andrewsykim @aojea Thanks for the great feedback. I think I've addressed everything now. PTAL. Test failures seem to be related to https://github.com/kubernetes/test-infra/issues/19910, still trying to figure out what's happening there.

robscott commented:
/retest

robscott commented:
/retest

andrewsykim commented on pkg/registry/discovery/endpointslice/strategy.go:
```diff
@@ -103,41 +103,55 @@ func (endpointSliceStrategy) AllowUnconditionalUpdate() bool {
-		endpointSlice.Endpoints[i].Conditions.Serving = nil
-		endpointSlice.Endpoints[i].Conditions.Terminating = nil
+	if !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName) {
+		for i := range endpointSlice.Endpoints {
```
I was wondering this question to myself as I was working on the `EndpointSliceTerminatingCondition` feature and I think it might be worthwhile to bring it up now. In v1.20 where most clusters will have both the `EndpointSliceTerminatingCondition` and `EndpointSliceNodeName` gates off, could it be problematic performance-wise that on CREATE and UPDATE we iterate over the entire endpoints list for every EndpointSlice two times to drop disabled fields? Or is this negligible given each EndpointSlice will typically not have more than 100 endpoints?

andrewsykim replied:
and even if it's negligible, is it worth optimizing this so we drop disabled fields in one loop?
```
if utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceTerminatingCondition) && utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName) {
    return
}

dropTerminating := !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceTerminatingCondition)
dropNodeName := !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName)
for i := range endpointSlice.Endpoints {
    if dropTerminating {
        endpointSlice.Endpoints[i].Conditions.Serving = nil
        endpointSlice.Endpoints[i].Conditions.Terminating = nil
    }

    if dropNodeName {
        endpointSlice.Endpoints[i].NodeName = nil
    }
}
```

andrewsykim replied:
same comment for Update

robscott replied:
Good call, updated to use that approach now.

andrewsykim commented on pkg/registry/discovery/endpointslice/strategy.go:
```diff
@@ -103,41 +103,56 @@ func (endpointSliceStrategy) AllowUnconditionalUpdate() bool {
+// dropDisabledFieldsOnUpdate will drop any disable fields that have not already
+// been set on the EndpointSlice.
+func dropDisabledFieldsOnUpdate(oldEPS, newEPS *discovery.EndpointSlice) {
+	dropNodeName := !utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName)
```
I think this will drop existing resources with the field already set when the feature gate is off, which I think is not desirable.

andrewsykim replied:
nvm ignore me, I misread :)

robscott replied:
I may not be understanding this comment correctly, but right below this line I set this to false if `NodeName` has been set on any endpoint:

```
	if dropNodeName {
		for _, ep := range oldEPS.Endpoints {
			if ep.NodeName != nil {
				dropNodeName = false
				break
			}
		}
	}
```

robscott replied:
I was about to link to the test that covered this and then I realized the gate was accidentally on in the test. That should be updated now and the corrected test does still pass.

andrewsykim commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -110,6 +120,17 @@ func validateEndpoints(endpoints []discovery.Endpoint, addrType discovery.Addres
 			}
 		}
 
+		if endpoint.NodeName != nil {
```
I never really understood why we check the gate in validation. I see three possible scenarios here:
1. the gate was on and we didn't drop it in strategy -- no need to check `allowNodeName` here since strategy allowed the field to be non-nil.
2. the gate is off and the field was dropped -- this check would be skipped since it's nil
3. the gate is off and on roll back a resource has the field set -- this is identical to 1. in terms of validation logic.

robscott replied:
Yeah I'm not sure if this really has any value since the strategy should cover everything. I have seen this pattern in a number of validation files though, so I'd lean towards leaving it in just in case we're missing anything.

thockin replied:
We should not be checking gates except rare cases

robscott replied:
Removed, thanks!

robscott commented:
/retest

thockin commented on pkg/controlplane/reconcilers/endpointsadapter_test.go:
```diff
@@ -232,7 +232,7 @@ func TestEndpointsAdapterUpdate(t *testing.T) {
 	endpoints4, _ := generateEndpointsAndSlice("foo", "testing", []int{80}, []string{"10.1.2.7", "10.1.2.8"})
 	_, epSlice4IP := generateEndpointsAndSlice("foo", "testing", []int{80}, []string{"10.1.2.7", "10.1.2.8"})
-	epSlice4IP.AddressType = discovery.AddressTypeIP
+	epSlice4IP.AddressType = discovery.AddressType("IP")
```
this could use a comment on why an "invalid" type is being used

robscott replied:
Updated, thanks!

thockin commented on pkg/controller/endpointslice/reconciler_test.go:
```diff
@@ -917,7 +917,7 @@ func TestReconcileEndpointSlicesReplaceDeprecated(t *testing.T) {
 
 	svc, endpointMeta := newServiceAndEndpointMeta("foo", namespace)
-	endpointMeta.AddressType = discovery.AddressTypeIP
+	endpointMeta.AddressType = discovery.AddressType("IP")
```
also a comment on why using an invalid type?

robscott replied:
Updated, thanks!

robscott committed:
Updates related to PR feedback

- Remove feature gate consideration from EndpointSlice validation
- Deprecate topology field, note that it will be removed in future
release
- Update kube-proxy to check for NodeName if feature gate is enabled
- Add comments indicating the feature gates that can be used to enable
alpha API fields
- Add comments explaining use of deprecated address type in tests

robscott commented on staging/src/k8s.io/api/discovery/v1beta1/types.go:
```diff
@@ -108,8 +102,13 @@ type Endpoint struct {
 	//   endpoint is located. This should match the corresponding node label.
 	// * topology.kubernetes.io/region: the value indicates the region where the
 	//   endpoint is located. This should match the corresponding node label.
+	// this field will be deprecated in an upcoming release.
```
Updated, thanks!

robscott commented on pkg/apis/discovery/validation/validation.go:
```diff
@@ -53,29 +52,44 @@ var (
 func ValidateEndpointSliceCreate(endpointSlice *discovery.EndpointSlice) field.ErrorList {
-	return ValidateEndpointSlice(endpointSlice, supportedAddressTypes)
+	// allow NodeName value if the feature gate is set.
+	allowNodeName := utilfeature.DefaultFeatureGate.Enabled(features.EndpointSliceNodeName)
```
Removed, thanks!

robscott commented:
@andrewsykim @thockin Thanks for the great feedback! I've pushed a commit that includes the following changes:

- Remove feature gate consideration from EndpointSlice validation
- Deprecate topology field, note that it will be removed in future release
- Update kube-proxy to check for NodeName first
- Add comments indicating the feature gates that can be used to enable alpha API fields
- Add comments explaining use of deprecated address type in tests

thockin reviewed (commented):
Thanks!

/lgtm
/approve

robscott commented:
/retest
